File: tools/reference-implementations/go/verify_vectors.go
package main

import (
	"encoding/hex"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strconv"
)

type doc struct {
	SpecVersion       string              `json:"spec_version"`
	CRC32CVectors     []crcVector         `json:"crc32c_vectors"`
	FrameVectors      []frameVector       `json:"frame_vectors"`
	SecureFrameVectors []secureFrameVector `json:"secure_frame_vectors"`
}

type crcVector struct {
	ID              string `json:"id"`
	InputHex        string `json:"input_hex"`
	ExpectedCRCHex  string `json:"expected_crc_hex"`
}

type frameInputs struct {
	Status    string `json:"status"`
	PID       uint32 `json:"pid"`
	Timestamp uint64 `json:"timestamp"`
	Nonce     uint64 `json:"nonce"`
	Payload   uint32 `json:"payload"`
}

type frameVector struct {
	ID                  string       `json:"id"`
	Kind                string       `json:"kind"`
	Inputs              *frameInputs `json:"inputs,omitempty"`
	ExpectedWireHex     string       `json:"expected_wire_hex"`
	WireHex             string       `json:"wire_hex"`
	ExpectedDecodeError *string      `json:"expected_decode_error"`
}

type secureFrameVector struct {
	ID                   string `json:"id"`
	Kind                 string `json:"kind"`
	KeyHex               string `json:"key_hex"`
	MasterKeyHex         string `json:"master_key_hex"`
	AgentKeyHex          string `json:"agent_key_hex"`
	DerivedAgentKeyHex   string `json:"derived_agent_key_hex"`
	AgentID              uint32 `json:"agent_id"`
	AgentPID             uint32 `json:"agent_pid"`
	IVRandomHex          string `json:"iv_random_hex"`
	IVCounter            uint32 `json:"iv_counter"`
	SessionSaltHex       string `json:"session_salt_hex"`
	PrefixIndex          uint32 `json:"prefix_index"`
	Epoch                uint64 `json:"epoch"`
	PlaintextHex         string `json:"plaintext_hex"`
	ExpectedWireHex      string `json:"expected_wire_hex"`
	ExpectedOKMHex       string `json:"expected_okm_hex"`
	ExpectedIVPrefixHex  string `json:"expected_iv_prefix_hex"`
}

func main() {
	defaultPath, _ := filepath.Abs("../../vlp-test-vectors.json")
	path := defaultPath
	if len(os.Args) > 1 {
		path = os.Args[1]
	}
	raw, err := os.ReadFile(path)
	if err != nil {
		die("read %s: %v", path, err)
	}
	var d doc
	if err := json.Unmarshal(raw, &d); err != nil {
		die("parse %s: %v", path, err)
	}

	var failures []string

	// ----- CRC -----
	for _, v := range d.CRC32CVectors {
		input, _ := hex.DecodeString(v.InputHex)
		got := CRC32C(input)
		want64, _ := strconv.ParseUint(v.ExpectedCRCHex, 16, 32)
		if uint64(got) != want64 {
			failures = append(failures, fmt.Sprintf("crc/%s: got %08x, want %08x", v.ID, got, want64))
		}
	}
	fmt.Printf("crc32c_vectors:        %d OK\n", len(d.CRC32CVectors))

	// ----- Frames -----
	roundtrip := 0
	errs := 0
	for _, v := range d.FrameVectors {
		switch v.Kind {
		case "encode_decode_roundtrip":
			if v.Inputs == nil {
				failures = append(failures, fmt.Sprintf("frame/%s: missing inputs", v.ID))
				continue
			}
			st, err := StatusFromName(v.Inputs.Status)
			if err != nil {
				failures = append(failures, fmt.Sprintf("frame/%s: %v", v.ID, err))
				continue
			}
			wire := Encode(st, v.Inputs.PID, v.Inputs.Timestamp, v.Inputs.Nonce, v.Inputs.Payload)
			if hex.EncodeToString(wire[:]) != v.ExpectedWireHex {
				failures = append(failures, fmt.Sprintf(
					"frame/%s: encode mismatch\n  got  %s\n  want %s",
					v.ID, hex.EncodeToString(wire[:]), v.ExpectedWireHex,
				))
				continue
			}
			rawBytes, _ := hex.DecodeString(v.ExpectedWireHex)
			frame, err := Decode(rawBytes)
			if err != nil {
				failures = append(failures, fmt.Sprintf("frame/%s: decode err %v", v.ID, err))
				continue
			}
			if frame.PID != v.Inputs.PID || frame.Timestamp != v.Inputs.Timestamp ||
				frame.Nonce != v.Inputs.Nonce || frame.Payload != v.Inputs.Payload {
				failures = append(failures, fmt.Sprintf("frame/%s: decoded fields differ", v.ID))
				continue
			}
			roundtrip++
		case "decode_error":
			wantErr := ""
			if v.ExpectedDecodeError != nil {
				wantErr = *v.ExpectedDecodeError
			}
			rawBytes, _ := hex.DecodeString(v.WireHex)
			_, err := Decode(rawBytes)
			if err == nil {
				failures = append(failures, fmt.Sprintf("frame/%s: expected error %s, got OK", v.ID, wantErr))
				continue
			}
			de, ok := err.(*DecodeError)
			if !ok || de.Kind != wantErr {
				failures = append(failures, fmt.Sprintf("frame/%s: expected %s, got %v", v.ID, wantErr, err))
				continue
			}
			errs++
		default:
			failures = append(failures, fmt.Sprintf("frame/%s: unknown kind %q", v.ID, v.Kind))
		}
	}
	fmt.Printf("frame_vectors:         %d round-trips, %d error vectors OK\n", roundtrip, errs)

	// ----- Secure -----
	secureCount := 0
	for _, v := range d.SecureFrameVectors {
		switch v.Kind {
		case "shared_key_seal":
			var key [32]byte
			var ivR [8]byte
			var pt [32]byte
			copyHex(key[:], v.KeyHex)
			copyHex(ivR[:], v.IVRandomHex)
			copyHex(pt[:], v.PlaintextHex)
			wire, err := EncodeShared(key, ivR, v.IVCounter, pt)
			if err != nil {
				failures = append(failures, fmt.Sprintf("secure/%s: %v", v.ID, err))
				continue
			}
			if hex.EncodeToString(wire[:]) != v.ExpectedWireHex {
				failures = append(failures, fmt.Sprintf("secure/%s: wire mismatch", v.ID))
				continue
			}
		case "master_key_seal":
			var master [32]byte
			var ivR [8]byte
			var pt [32]byte
			copyHex(master[:], v.MasterKeyHex)
			copyHex(ivR[:], v.IVRandomHex)
			copyHex(pt[:], v.PlaintextHex)
			wire, derived, err := EncodeMaster(master, v.AgentPID, ivR, v.IVCounter, pt)
			if err != nil {
				failures = append(failures, fmt.Sprintf("secure/%s: %v", v.ID, err))
				continue
			}
			if hex.EncodeToString(derived[:]) != v.DerivedAgentKeyHex {
				failures = append(failures, fmt.Sprintf("secure/%s: agent-key derivation mismatch", v.ID))
				continue
			}
			if hex.EncodeToString(wire[:]) != v.ExpectedWireHex {
				failures = append(failures, fmt.Sprintf("secure/%s: wire mismatch", v.ID))
				continue
			}
		case "kdf_agent_key":
			var master [32]byte
			copyHex(master[:], v.MasterKeyHex)
			key := DeriveAgentKey(master, v.AgentID)
			if hex.EncodeToString(key[:]) != v.ExpectedOKMHex {
				failures = append(failures, fmt.Sprintf("secure/%s: HKDF mismatch", v.ID))
				continue
			}
		case "kdf_iv_prefix":
			var salt [16]byte
			copyHex(salt[:], v.SessionSaltHex)
			iv := DeriveIVPrefix(salt, v.PrefixIndex)
			if hex.EncodeToString(iv[:]) != v.ExpectedIVPrefixHex {
				failures = append(failures, fmt.Sprintf("secure/%s: HKDF mismatch", v.ID))
				continue
			}
		case "kdf_epoch_key":
			var agent [32]byte
			copyHex(agent[:], v.AgentKeyHex)
			key := DeriveEpochKey(agent, v.Epoch)
			if hex.EncodeToString(key[:]) != v.ExpectedOKMHex {
				failures = append(failures, fmt.Sprintf("secure/%s: HKDF mismatch", v.ID))
				continue
			}
		default:
			failures = append(failures, fmt.Sprintf("secure/%s: unknown kind %q", v.ID, v.Kind))
			continue
		}
		secureCount++
	}
	fmt.Printf("secure_frame_vectors:  %d OK\n", secureCount)

	if len(failures) > 0 {
		fmt.Println()
		fmt.Println("FAILED:")
		for _, f := range failures {
			fmt.Printf("  - %s\n", f)
		}
		os.Exit(1)
	}
	fmt.Println()
	fmt.Println("ALL VECTORS PASSED")
}

func copyHex(dst []byte, src string) {
	b, err := hex.DecodeString(src)
	if err != nil || len(b) != len(dst) {
		die("hex %q: expected %d bytes, got %d (%v)", src, len(dst), len(b), err)
	}
	copy(dst, b)
}

func die(format string, args ...any) {
	fmt.Fprintf(os.Stderr, format+"\n", args...)
	os.Exit(1)
}

File: tools/reference-implementations/python/verify_vectors.py
#!/usr/bin/env python3
"""Run the VLP conformance suite against the Python reference implementation.

Usage:

    python3 verify_vectors.py [PATH_TO_JSON]

Default JSON path is `../../vlp-test-vectors.json` (relative to this script).

Exits 0 on full pass. Secure-frame vectors require the third-party
``cryptography`` package — they are skipped (with a warning) if it is
not installed.
"""

from __future__ import annotations

import json
import sys
from pathlib import Path

import vlp


def main(argv: list[str]) -> int:
    here = Path(__file__).resolve().parent
    default = here.parent.parent / "vlp-test-vectors.json"
    path = Path(argv[1]) if len(argv) > 1 else default
    doc = json.loads(path.read_text())

    failures: list[str] = []

    # ----- CRC -----
    for v in doc["crc32c_vectors"]:
        got = vlp.crc32c(bytes.fromhex(v["input_hex"]))
        want = int(v["expected_crc_hex"], 16)
        if got != want:
            failures.append(f"crc/{v['id']}: got {got:08x}, want {want:08x}")
    print(f"crc32c_vectors:        {len(doc['crc32c_vectors'])} OK")

    # ----- Frames -----
    roundtrip = 0
    error = 0
    for v in doc["frame_vectors"]:
        kind = v["kind"]
        if kind == "encode_decode_roundtrip":
            i = v["inputs"]
            wire = vlp.encode(i["status"], i["pid"], i["timestamp"],
                              i["nonce"], i["payload"])
            if wire.hex() != v["expected_wire_hex"]:
                failures.append(
                    f"frame/{v['id']}: encode mismatch\n"
                    f"  got  {wire.hex()}\n"
                    f"  want {v['expected_wire_hex']}"
                )
                continue
            decoded = vlp.decode(bytes.fromhex(v["expected_wire_hex"]))
            if decoded.pid != i["pid"] or decoded.timestamp != i["timestamp"] \
                    or decoded.nonce != i["nonce"] or decoded.payload != i["payload"]:
                failures.append(f"frame/{v['id']}: decode roundtrip mismatch")
                continue
            roundtrip += 1
        elif kind == "decode_error":
            try:
                vlp.decode(bytes.fromhex(v["wire_hex"]))
                failures.append(f"frame/{v['id']}: expected error, got OK")
                continue
            except vlp.DecodeError as e:
                if e.kind != v["expected_decode_error"]:
                    failures.append(
                        f"frame/{v['id']}: expected {v['expected_decode_error']}, "
                        f"got {e.kind} ({e})"
                    )
                    continue
                error += 1
        else:
            failures.append(f"frame/{v['id']}: unknown kind {kind}")
    print(f"frame_vectors:         {roundtrip} round-trips, {error} error vectors OK")

    # ----- Secure frames -----
    secure_done = 0
    secure_skipped = 0
    try:
        import vlp_secure  # noqa: F401
        have_secure = True
        # Probe AEAD availability separately.
        try:
            vlp_secure._aead()
        except ImportError:
            print("  (skip) cryptography package not installed — secure_frame_vectors skipped.")
            print("         Install with: pip install cryptography")
            have_secure = False
    except ImportError:
        have_secure = False

    if have_secure:
        for v in doc["secure_frame_vectors"]:
            kind = v["kind"]
            try:
                if kind == "shared_key_seal":
                    wire = vlp_secure.encode_shared(
                        bytes.fromhex(v["key_hex"]),
                        bytes.fromhex(v["iv_random_hex"]),
                        v["iv_counter"],
                        bytes.fromhex(v["plaintext_hex"]),
                    )
                    if wire.hex() != v["expected_wire_hex"]:
                        failures.append(f"secure/{v['id']}: wire mismatch")
                        continue
                elif kind == "master_key_seal":
                    derived = vlp_secure.derive_agent_key(
                        bytes.fromhex(v["master_key_hex"]),
                        v["agent_pid"],
                    )
                    if derived.hex() != v["derived_agent_key_hex"]:
                        failures.append(f"secure/{v['id']}: agent-key derivation mismatch")
                        continue
                    wire = vlp_secure.encode_master(
                        bytes.fromhex(v["master_key_hex"]),
                        v["agent_pid"],
                        bytes.fromhex(v["iv_random_hex"]),
                        v["iv_counter"],
                        bytes.fromhex(v["plaintext_hex"]),
                    )
                    if wire.hex() != v["expected_wire_hex"]:
                        failures.append(f"secure/{v['id']}: wire mismatch")
                        continue
                elif kind == "kdf_agent_key":
                    derived = vlp_secure.derive_agent_key(
                        bytes.fromhex(v["master_key_hex"]),
                        v["agent_id"],
                    )
                    if derived.hex() != v["expected_okm_hex"]:
                        failures.append(f"secure/{v['id']}: HKDF mismatch")
                        continue
                elif kind == "kdf_iv_prefix":
                    derived = vlp_secure.derive_iv_prefix(
                        bytes.fromhex(v["session_salt_hex"]),
                        v["prefix_index"],
                    )
                    if derived.hex() != v["expected_iv_prefix_hex"]:
                        failures.append(f"secure/{v['id']}: HKDF mismatch")
                        continue
                elif kind == "kdf_epoch_key":
                    derived = vlp_secure.derive_epoch_key(
                        bytes.fromhex(v["agent_key_hex"]),
                        v["epoch"],
                    )
                    if derived.hex() != v["expected_okm_hex"]:
                        failures.append(f"secure/{v['id']}: HKDF mismatch")
                        continue
                else:
                    failures.append(f"secure/{v['id']}: unknown kind {kind}")
                    continue
                secure_done += 1
            except Exception as e:
                failures.append(f"secure/{v['id']}: exception {e!r}")
        print(f"secure_frame_vectors:  {secure_done} OK")
    else:
        secure_skipped = len(doc["secure_frame_vectors"])
        print(f"secure_frame_vectors:  {secure_skipped} skipped (no cryptography pkg)")

    if failures:
        print()
        print("FAILED:")
        for f in failures:
            print(f"  - {f}")
        return 1

    print()
    print("ALL VECTORS PASSED")
    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv))
